modules/core/include/opencv2/core/utility.hpp

```diff
@@ -513,30 +513,36 @@ class CV_EXPORTS AutoLock
     AutoLock& operator = (const AutoLock&);
 };
 
+// TLS interface
 class CV_EXPORTS TLSDataContainer
 {
-private:
-    int key_;
 protected:
-    TLSDataContainer();
-    virtual ~TLSDataContainer();
-public:
-    virtual void* createDataInstance() const = 0;
-    virtual void deleteDataInstance(void* data) const = 0;
+    TLSDataContainer() { key = (size_t)-1; };
+    virtual ~TLSDataContainer() {};
 
+    void  getKey();
     void* getData() const;
+    void  releaseKey();
+
+private:
```

Comment:
Right, "public" access is not required anymore. It will be useful to remove it for future releases with ABI breaks.
Let mark this something like:
```
#ifdef OPENCV_ABI_COMPATIBILITY
public:
#else
private:
#endif
```

Same trick can be done with "int->size_t" for `key`.